src/slip0010/ckd.rs

```diff
@@ -0,0 +1,284 @@
+//! Utilities for Child Key Derivation (CKD) according to SLIP-0010.
+
+use crate::{
+    errors::CallerError,
+    utils::{bn_to_scalar, CurvePoint},
+};
+use generic_array::{
+    typenum::{U32, U64},
+    GenericArray,
+};
+use hmac::Mac;
+use k256::Scalar;
+use libpaillier::unknown_order::BigNumber;
+
+type HmacSha512 = hmac::Hmac<sha2::Sha512>;
+
+/// Represents the input to the CKD function.
+#[derive(Debug, Clone)]
+pub struct CKDInput {
+    private_key: Option<Scalar>,
+    public_key: Vec<u8>,
+    chain_code: [u8; 32],
+    index: u32,
+}
+
+/// Represents the output of the CKD function.
+#[derive(Debug, Clone)]
+pub struct CKDOutput {
+    /// The chain code.
+    pub chain_code: [u8; 32],
+    /// The scalar.
+    pub private_key: Scalar,
+}
+
+/// Represents the input to the master key derivation function.
+#[derive(Debug, Clone)]
+pub struct MasterKeyInput {
+    seed: Vec<u8>,
+    curve: String,
+}
+
+impl MasterKeyInput {
+    /// Create a new MasterKeyInput.
+    pub fn new(seed: &[u8], curve: String) -> Result<Self, CallerError> {
+        if curve != "Bitcoin seed" {
+            Err(CallerError::WrongCurve)
+        } else {
+            Ok(Self {
+                seed: seed.to_vec(),
+                curve,
+            })
+        }
+    }
+
+    /// Derive master key from seed.
+    pub fn derive_master_key(&self) -> CKDOutput {
+        let hmac = HmacSha512::new_from_slice(self.curve.as_bytes())
+            .expect("this never fails: hmac can handle keys of any size");
+        let i = hmac
+            .clone()
+            .chain_update(self.seed.clone())
+            .finalize()
+            .into_bytes();
+        let (i_left, i_right) = split_into_two_halfes(&i);
```

Comment:
It is possible for this step to fail with SLIP 10. I know we do not really need this code for the tss library but since we have it I think it would be best if we implemented the whole thing. It should be a very minor change.
<img width="745" alt="Screenshot 2024-10-06 at 1 33 32 PM" src="https://github.com/user-attachments/assets/38167a2b-245f-48ba-a37d-b0e1054e67c0">

Reply:
done